### PR TITLE
Framework detection

### PR DESCRIPTION
**Pull request by nturinski**
Related to https://github.com/microsoft/vscode-azurestaticwebapps/issues/465

This is our first pass at trying to detect what framework the user's project is.  It's based heavily on Oryx's implementation as seen here: https://github.com/microsoft/Oryx/tree/main/src/Detector/Node

Because I have no idea about the accuracy of this detector, we're just going to compare our results to what the user selects at the `BuildPresetStep`.

Current limitations:
- If the folder is in a subfolder (which seems likely), it will not detect
- Does not look for the app folder if it is not the root directory
- Can return multiple frameworks since it's doing detection via package.json parsing

### CODE REVIEW

File: src/detectors/node/NodeDetector.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    frameworks: FrameworkInfo[],
+    hasLernaJsonFile: boolean,
+    hasLageConfigJSFile: boolean,
+    lernaNpmClient: string | undefined,
+    hasYarnrcYmlFile: boolean,
+    isYarnLockFileValidYamlFormat: boolean,
+}
+
+
+export class NodeDetector {
```
Comment (alexweininger) [inline src/detectors/node/NodeDetector.ts:41]: A comment somewhere prominent in this file explaining that this code is modeled after Oryx, and a link to the Oryx detector would be good.

And now that I think about it, probably the NodeConstants.ts file also.

**Review by alexweininger — COMMENTED** [review]
Two things:

1. Talked offline about updating this PR to be more virtual workspace and github.dev friendly. For example, reducing usages of `uri.fsPath`.

	More info and resources:
	* https://code.visualstudio.com/api/extension-guides/web-extensions#migrate-extension-with-code
	* https://github.com/microsoft/vscode/wiki/Virtual-Workspaces#review-that-the-code-is-ready-for-virtual-resources

2. I see that git and GitHub think that the test projects are git submodules. I am not familiar with git submodules and just want to make sure this won't cause issues. We could probably exclude the nested .git folders to prevent this.

image

File: src/detectors/node/NodeDetector.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    version: string
+}
+
+export type DetectorResults = {
+    platform: string | undefined,
+    platformVersion: string | undefined,
+    appDirectory: string | undefined,
+    frameworks: FrameworkInfo[],
+    hasLernaJsonFile: boolean,
+    hasLageConfigJSFile: boolean,
```
[the remark this answers is not in the document]
Comment (nturinski) [reply src/detectors/node/NodeDetector.ts:34]: This actually refers to this: https://microsoft.github.io/lage/guide/

Seems like it's a tool to run your build/test pipeline more asynchronously.

**Comment by nturinski** [comment]
Closing because I need to do a clean push of the branch so that the test projects are not interpreted as submodules.  As it turns out, it is causing problems because it's not actually copying the content into our repo which means the tests have nothing to run against.

File: src/detectors/node/NodeDetector.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexweininger) [reply src/detectors/node/NodeDetector.ts:37]: Haha makes more sense now. I was wondering what a large config file might be... 😂

**Review by alexweininger — APPROVED** [review]

**Comment by anthonychu** [comment]
@nturinski This is cool. Can we get a demo of this? /cc @simonaco @fiveisprime

**Comment by alexweininger** [comment]
@anthonychu 

Currently we aren't actually doing anything with the detection result. We wanted to verify the detection logic is accurate enough to trust in terms of setting things up automatically for users down the road, or filling in default values in prompts, etc. 

Once we are actually using the detector result, I'll be glad to give a demo. 😄

**Pull request merged**